File: core/src/me/shreyasr/cota/system/CollisionSystem.java
package me.shreyasr.cota.system;

import com.badlogic.ashley.core.Entity;
import com.badlogic.ashley.core.Family;
import com.badlogic.ashley.systems.SortedIteratingSystem;
import me.shreyasr.cota.component.StateDataComponent;
import me.shreyasr.cota.util.EntityIdComparator;
import me.shreyasr.cota.util.Rectangle;

public class CollisionSystem extends SortedIteratingSystem {

    public CollisionSystem(int priority) {
        super(Family.all(StateDataComponent.class).get(), new EntityIdComparator() ,priority);
    }

    @Override
    protected void processEntity(Entity entity, float deltaTime) {
        for(Entity e : getEntities()){
            if(entity == e) continue;
            Rectangle otherRect = e.getComponent(StateDataComponent.class).rect();
            Rectangle meRect = entity.getComponent(StateDataComponent.class).rect();
            if(otherRect.isIn(meRect)){
                boolean otherGreaterX = otherRect.pos.x > meRect.pos.x;
                boolean otherGreaterY = otherRect.pos.y > meRect.pos.y;
                float targetX = otherGreaterX ? (otherRect.left() + meRect.right())/2 : (meRect.left() + otherRect.right())/2;
                float targetY = otherGreaterY ? (otherRect.down() + meRect.up())/2 : (meRect.up() + otherRect.down())/2;
                otherRect.pos.x = otherGreaterX ? targetX + otherRect.size.x/2 : targetX - otherRect.size.x/2;
                meRect.pos.x = !otherGreaterX ? targetX + otherRect.size.x/2 : targetX - otherRect.size.x/2;
                otherRect.pos.y = otherGreaterY ? targetY + otherRect.size.y/2 : targetX - otherRect.size.y/2;
                meRect.pos.x = !otherGreaterY ? targetY + otherRect.size.y/2 : targetX - otherRect.size.y/2;
            }
        }

    }
}

File: core/src/me/shreyasr/cota/system/CollisionSystem.scala
package me.shreyasr.cota.system

import com.badlogic.ashley.core.{Entity, Family}
import com.badlogic.ashley.systems.SortedIteratingSystem
import com.badlogic.gdx.graphics.Color
import me.shreyasr.cota._
import me.shreyasr.cota.component._
import me.shreyasr.cota.component.TypeComponent.Bullet
import me.shreyasr.cota.component.TypeComponent.Melee
import me.shreyasr.cota.util.EntityIdComparator

import scala.collection.JavaConverters._

class CollisionSystem(priority: Int)
  extends SortedIteratingSystem(Family.all(classOf[StateDataComponent]).get, EntityIdComparator, priority) {

  def processEntity(me: Entity, deltaTime: Float) {
    if (me.is[Bullet]){
      me.get[HealthComponent].takeDamage(1)
      if(me.get[HealthComponent].health<0) {
        getEngine.removeEntity(me)
      }
      return
    }
    if (me.is[Melee]){
      me.get[HealthComponent].takeDamage(1)
      if(me.get[HealthComponent].health<0) {
        getEngine.removeEntity(me)
      }
      return
    }
    getEntities.asScala
      .filterNot(_ == me)
      .filterNot(_.getOpt[OwnerIdComponent].exists(_.id == me.id))
      .foreach(other => {
        val otherState = other.get[StateDataComponent]
        val myState = me.get[StateDataComponent]
        if (otherState.rect.isIn(myState.rect, myState.pos, otherState.pos)) {
          if(other.is[Bullet]){
            me.get[HealthComponent].takeDamage(other.get[DamageComponent].dam)
            getEngine.removeEntity(other)
          }else if(other.is[Melee]){
            me.get[HealthComponent].takeDamage(other.get[DamageComponent].dam)
          }
          other.get[RenderDataComponent].color = Color.RED
          me.get[RenderDataComponent].color = Color.RED
        }
      })
  }
}
